### PR TITLE
Admin API

### PR DESCRIPTION
**Pull request by zshipko**
**- Summary**

This pull request introduces the Admin API for managing gotrue instances. Additionally, an `admin` command has been added to the command-line application for performing some of these actions. This also includes merging work from the `audience_support` branch and a few other changes outlined below. 

- Merged support for audience claim from `audience_support` branch (#14)
- Admin API for listing, creating, deleting and updating users (#30) 
  - Added `IsSuperAdmin` to the User model
  - Added `Role` to the User model, making it possible for each user to only have one role per audience
- Made confirmation flow optional by adding Configuration.Mailer.Autoconfirm option (#28)
- Added Configuration.JWT.DefaultGroupName to determine the role each user should be assigned at signup.
- Added email validation using [checkmail](https://github.com/badoux/checkmail)

**- Test plan**

I've continued to write tests for new features, and will continue to write more tests for edge cases if time permits. I also created the beginnings of a [test website](https://github.com/zshipko/gotrue-example-website) for manual testing. All Go tests are passing (using both mongo and sqlite3) and everything seems to be working correctly with the new changes.

**- Description for the changelog**

Merge audience support and added admin API. 

**- A picture of a cute animal (not mandatory but encouraged)**

:-P

### CODE REVIEW

File: api/admin_test.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+	resp := w.Result()
+
+	if resp.StatusCode == 200 {
+		t.Log(resp)
+		t.Fail()
+	}
+}
+
+func makeSuperAdmin(req *http.Request, api *API, email string, t *testing.T) (context.Context, *httptest.ResponseRecorder) {
```
Comment (rybit) [inline api/admin_test.go]: tiny thing (as in feel free to ignore) but require/assert take the `t *testing.T` as the first param. I tend to try to follow that convention, but def not necessary.

File: api/admin.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+	if params.User.Aud == "" {
+		params.User.Aud = aud
+	}
+
+	user, err := api.db.FindUserByEmailAndAudience(params.User.Email, params.User.Aud)
+	if err != nil {
+		if user, err = api.db.FindUserByID(params.User.ID); err != nil && requireUser {
+			if models.IsNotFoundError(err) {
+				NotFoundError(w, err.Error())
```
Comment (rybit) [inline api/admin.go]: more places `errors.Wrap` would be good.

File: api/admin_test.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+		}
+	}
+
+	u, err = models.NewUser(email, "test", api.config.JWT.Aud, nil)
+	if err != nil {
+		t.Error(err)
+	}
+
+	u.IsSuperAdmin = true
+	api.db.CreateUser(u)
```
Comment (rybit) [inline api/admin_test.go]: you'll want to check the return here.
`require.NoError(t, api.db.CreateUser(u))`

File: api/admin_test.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		t.Error(err)
+	}
+
+	if resp.StatusCode != 200 {
+		t.Log(resp)
+		t.Fail()
+	}
+
+	if len(data["users"].([]interface{})) < 1 {
```
Comment (rybit) [inline api/admin_test.go]: I think you should return and at least check a user here. It could be that you return a list of empty users, which would be wrong.

File: api/admin_test.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+	// Setup request
+	req := httptest.NewRequest("POST", "/admin/user", &buffer)
+
+	// Setup response recorder with super admin privileges
+	ctx, w := makeSuperAdmin(req, api, "test@example.com", t)
+
+	api.adminUserCreate(ctx, w, req)
+
+	resp := w.Result()
```
Comment (rybit) [inline api/admin_test.go]: also iirc `ResponseRecorder` has a `w.Code` and a way to easily get the body as []byte. It would mean you could do 
```
assert.Equal(t, http.StatusOK, w.Code)
payload := new(...)
require.NoError(t, json.Unmarshal(w.Body.Bytes(), payload))
```

File: cmd/admin_cmd.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+func getAudience(c *conf.Configuration) string {
+	if audience == "" {
+		return c.JWT.Aud
+	}
+
+	return audience
+}
+
+func init() {
```
Comment (rybit) [inline cmd/admin_cmd.go]: personally, I prefer not using `init()`, I would rather have a method `AdminCmd() *cobra.Command` that did all of this and then had a locally scoped `adminCmd`. 

Feel free to ignore.

**Comment by zshipko** [comment]
@rybit I've gone through and fixed all the issues you pointed out. Let me know if you think any more changes are needed, otherwise I will begin working on the external providers task next.

File: api/admin_test.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+	ts.API.adminUsers(ctx, w, req)
+
+	resp := w.Result()
+	assert.Equal(ts.T(), resp.StatusCode, 401)
+}
+
+func (ts *AdminTestSuite) makeSuperAdmin(req *http.Request, email string) (context.Context, *httptest.ResponseRecorder) {
+	// Cleanup existing user
+	u, err := ts.API.db.FindUserByEmailAndAudience(email, ts.API.config.JWT.Aud)
+	if err == nil {
```
Comment (rybit) [inline api/admin_test.go]: instead of doing the nil check you can just use the `NoError` method

File: api/admin_test.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+	// Setup request
+	req := httptest.NewRequest("GET", "/admin/user", &buffer)
+
+	// Setup response recorder with super admin privileges
+	ctx, w := ts.makeSuperAdmin(req, "test@example.com")
+
+	ts.API.adminUserGet(ctx, w, req)
+
+	resp := w.Result()
```
Comment (rybit) [inline api/admin_test.go]: instead of pulling the result of the ResponseRecorder you can directly use its buffer and return codes.

File: api/signup.go [hunk 9]
```diff
@@ -53,16 +57,30 @@ func (a *API) Signup(ctx context.Context, w http.ResponseWriter, r *http.Request
 		}
 	}
 
-	if err := a.mailer.ConfirmationMail(user); err != nil {
-		InternalServerError(w, fmt.Sprintf("Error sending confirmation mail: %v", err))
-		return
+	if err = mailer.ValidateEmail(params.Email); err != nil {
+		if !a.config.Testing {
```
Comment (rybit) [inline api/signup.go]: Ideally we shouldn't have to have a `config.Testing` flag.

File: api/signup_test.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	ctx := req.Context()
+
+	ts.API.Signup(ctx, w, req)
+
+	resp := w.Result()
+	assert.Equal(ts.T(), resp.StatusCode, 200)
+
+	data := make(map[string]interface{})
+	require.NoError(ts.T(), json.NewDecoder(resp.Body).Decode(&data))
+	assert.Equal(ts.T(), data["email"], "test@example.com")
```
Comment (rybit) [inline api/signup_test.go]: I would also check the size of the map returned. So that you know what you expect to come back.
